### ISSUE TITLE
New rule to flag complex argument expression values

### ISSUE THREAD

**Issue by henrikbarium** [detekt/detekt#4592]
## Expected Behavior of the rule

The new rule would detect and flag function/constructor calls where complex expressions are used as argument values. It would accept literal values, variable references, and possibly simple expressions such as property access.

```kotlin
// Assuming this function (plus some other things which I leave out for brevity)
fun processCandidateForMission(
    birthName: String,
    dateOfBirth: LocalDate,
    fitessForPurpose: Float,
    willingToParticipate: Boolean,
    relatedSkills: List<Skill>
): Unit = TODO()

// 1) This call should be flagged as an error - complex expressions used as argument values
processCandidateForMission(
    currentUser.name,
    LocalDate.of(2000, 1, 1),
    calculateFitness(currentUser),
    if (compensation > 1000) true else Random.nextBoolean(),
    trendySkills.filter { it.level >= 5 }.sortedBy { it.name }.asReversed()
        .map { it.copy(name = it.name.uppercase().trim()) }
)

// 2) This call is ok - complex expressions extracted as separate variables which are then used as argument values
val dateOfBirth = LocalDate.of(2000, 1, 1)
val fitessForPurpose = calculateFitness(currentUser)
val willingToParticipate = if (compensation > 1000) true else Random.nextBoolean()
val relatedSkills = trendySkills.filter { it.level >= 5 }.sortedBy { it.name }.asReversed()
    .map { it.copy(name = it.name.uppercase().trim()) }
processCandidateForMission(currentUser.name, dateOfBirth, fitessForPurpose, willingToParticipate, relatedSkills)
```

In example (2) one might argue that `LocalDate.of(2000, 1 ,1)` is "less complex" than `calculateFitness(currentUser)` because it is a well known, side-effect free "date literal". So perhaps it would be ok not to extract. Unless there is a clear way to determine which cases are safe, maintaining such a whitelist could be burdensome, however.

## Context
The idea behind the suggestion is to make the code easier to understand and debug.

- Easier to understand as each statement does one (or at least fewer) things. When we get to the `processCandidateForMission` call, we can focus on that call itself and not have to analyze a lot of semi-complex expressions.
    - If one of the expressions also have side-effects in addition, I think this becomes even more important. Having a side-effect "hidden" in an argument value in a call to another function is very easy to miss.
- Easier to debug because the result of each "complex" expression is assigned to a variable which is easy to inspect. We can put a breakpoint just before calling `processCandidateForMission` and see what values will be sent to it.

This is somewhat related to named arguments as use of named arguments could be said to make the code easier to use. However, at least personally, I still find separate variables easier to read. Named arguments also does not make the code easier to inspect in a debugger.

I suspect that there may be several edge cases that I have not thought of, so maybe the rule would work slightly different than what I describe.

I also understand that what is right here may be subjective, but such a rule would bring value to me and my team at least.

**Comment by cortinico** [comment]
> // 2) This call is ok - complex expressions extracted as separate variables which are then used as argument values

I'm not sure I fully agree with this. Forcing export to expression would force users to convert a single function call to a block. Moreover, I think that named argumetns can actually tackle this scenario in an elegant way. 

To make an example, I prefer this:

```
fun myFunction(currentUser: User) = processCandidateForMission(
    birthName = currentUser.name,
    dateOfBirth = LocalDate.of(2000, 1, 1)
    fitessForPurpose = calculateFitness(currentUser)
    willingToParticipate = if (compensation > 1000) true else Random.nextBoolean()
    relatedSkills = trendySkills.filter { it.level >= 5 }.sortedBy { it.name }.asReversed()
)
```

to this:

```
fun myFunction(currentUser: User) : Int {
    val dateOfBirth = LocalDate.of(2000, 1, 1)
    val fitessForPurpose = calculateFitness(currentUser)
    val willingToParticipate = if (compensation > 1000) true else Random.nextBoolean()
    val relatedSkills = trendySkills.filter { it.level >= 5 }.sortedBy { it.name }.asReversed()
        .map { it.copy(name = it.name.uppercase().trim()) }
    processCandidateForMission(currentUser.name, dateOfBirth, fitessForPurpose, 
    	willingToParticipate, relatedSkills)
}
```

**Comment by henrikbarium** [comment]
I fully understand your position and I guess this is a case which is a bit subjective. (That is also a reason why it could work well as an optional setting for the rule.)

As for the two examples, I would often prefer the second. But it depends on the complexity and "conceptual weight" of the argument values. If the arguments are just literals or trivial expressions it is fine to do as in example 1. But if the arguments are calls to other non-trivial functions (which may even have side-effects) I think example 2 is clearer.

To me example 2 also reads more naturally. The call to `processCandidateForMission` comes last in the source code just as it does during execution. In the first example, all other lines are executed before the first. This becomes even more weird if some of the nested function calls are wrapped too.

And as I said in the original message, example 2 is much easier to work with in a debugger than example 1. Both because the execution order as just mentioned, but also because there are local variables with intermediate values to inspect.

But again, I understand that everyone will not agree. :-)

**Comment by henrikbarium** [comment]
As a side note, I think it is unfortunate (but acceptable) to have to turn a function with expression body to a block body. I wish Kotlin went the same way as Scala with just treating the last expression of a function as the return value. That is what Kotlin does in other blocks anyway. :-)

**Comment by cortinico** [comment]
> But again, I understand that everyone will not agree. :-)

It sounds like use case for a custom rule maybe.
Happy to hear other's opinion on this as well

**Comment by BraisGabin** [comment]
I agree with @cortinico here.

The problem that I see here is what "complex arguments expression values" means. What is a complex argument and what is not? That could (ironically) blow up the complexity of this rule. That would be a lot of work to maintain so I don't hink this rule should be in the default rule set of detekt. But that could make a really good custom rule.

**Comment by henrikbarium** [comment]
Ok, fair enough. I can see that it is a rule that would have to be "just right" - and that would probably be subjective.
Thanks for providing feedback!
